### DISCUSSION TITLE
Why were pymodbus.utilities 's checkCRC, computeCRC functions removed in pymodbus 3.6.6 when they were working in 3.6.5? Is there a better recommendation for what I should use for CRC?

### DISCUSSION THREAD

**Discussion by temataro** [pymodbus-dev/pymodbus#2178] [Help] [answered]
My code broke after updating to 3.6.6 and I was wondering if it wasn't best practice to use those functions anymore. And if so, what I should be using instead. Thanks!

**Comment by janiversen** [comment, marked as answer]
These function are not removed but moved, the RTU framer depends on the CRC check.


The functions are considered internal and not part of the API, and I am not sure I understand why you need to use them ?

**Reply by temataro** [reply]
Hi. Thank you for your reply!

The extent of my use case with pymodbus is to communicate with an RTU relay that required me to append the CRC bits to the end of my message. (The documentation for the product itself has [example scripts](https://www.waveshare.com/wiki/Modbus_RTU_Relay#Demo_codes) that implemented its own CRC checksum on the 6 bytes of instruction to pass onto the relay module).

I opted to find a module that would just do that for me and landed here.
That allowed me to reduce my code to just these lines:
```python
def append_crc(cmd: list[int]) -> list[int]:
  """Takes a 6 byte instruction in the form of a 6 element list and appends crc information to it."""
  crc = pymodbus.utilities.computeCRC(cmd)
  cmd.append(crc >> 8)
  cmd.append(crc & 0xff)

  return cmd
```
This is the way I'd been using it so far but if there's any other alternative to doing my own CRC handling, I'd love to hear about that too.
Thanks for all the hard work.

**Comment by janiversen** [comment]
you are basically doing your best not to use the features of pymodbus 😄 

Pymodbus offers a high level API to read register, coils etc, so really no need to create modbus packets. Take a look at e.g. our example simple_sync_client.py

Your way of doing it, still works with the newest pymodbus version, you just need to adapt your code, as is basically the case with any library update.

**Reply by janiversen** [reply]
The advantage of using pymodbus API, is that it takes care of connection/reconnect/error handling and a number of other things that can happen within the modbus protocol.

**Reply by temataro** [reply]
ha! I suppose I can just use other functions in the pymodbus API instead of trying to emulate whatever the manufacturer did in their implementation. I don't even need to close this thread by marking my solution since the answer was to ... just use more of pymodbus' features. I'll have a look at [`simple_sync_client.py`](https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/simple_sync_client.py) again and use that as my base. Thank you again.

**Comment by temataro** [comment]
Look at [`simple_sync_client.py`](https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/simple_sync_client.py) for a more complete implementation of RTU communication instead of emulating it with pymodbus.utilities.computeCRC and other convoluted methods. That function (along with other CRC related functions are no longer in pymodbus.utilities as of 3.6.6).
